include/mbgl/style/expression/match.hpp

```diff
@@ -0,0 +1,277 @@
+#pragma once
+
+#include <mbgl/style/expression/expression.hpp>
+#include <mbgl/style/expression/parsing_context.hpp>
+#include <mbgl/style/conversion.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+using InputType = variant<int64_t, std::string>;
+using MatchKey = variant<InputType, std::vector<InputType>>;
+
+template <typename T>
```

Comment:
Seems like we need to make a decision. We definitely want to support:

```
["match", input: T, (label: T | Array<T>, expression: R)..., otherwise: R] (labels must be literals)

```

for `T ∈ {number, string, boolean}`. Do we want to support:

```
["match", input: Value, (label: number | string | boolean | Array<number | string | boolean>, expression: R)..., otherwise: R] (labels must be literals)
```